lib/synapse/config_generator/README.md

```diff
@@ -0,0 +1,74 @@
+## ConfigGenerator Classes
+
+Generators are the piece of Synapse that react to changes in service
+registrations and actually reflect those changes in local state.
+Generators should conform to the interface specified by `BaseGenerator` and
+when your generator has received an update from synapse via `update_config` it
+should sync the watcher state with the external configuration (e.g. HAProxy
+state)
+
+```ruby
+require "synapse/config_generator/base"
+
+class Synapse::ConfigGenerator
+  class MyGenerator < BaseGenerator
+    # The generator name is used to find service specific
+    # configuration in the service watchers. When supplying
+    # per service config, use this as the key
+    NAME = 'my_generator'.freeze
+
+    def initialize(opts = {})
+        # Process and validate any options specified in the dedicated section
+        # for this config generator, given as the `opts` hash. You may omit
+        # this method, or you can declare your own, but remember to invoke
+        # the parent initializer
+        super(opts)
+    end
+
+    def update_config(watchers)
+      # synapse will call this method whenever watcher state changes with the
+      # watcher state. You should reflect that state in the local config state
+    end
+
+    def tick
+      # Called every loop of the main Synapse loop regardless of watcher
+      # changes (roughly ~1s). You can use this to rate limit how often your
+      # config generator actually reconfigures external services (e.g. HAProxy
+      # may need to rate limit reloads as those can be disruptive to in
+      # flight connections
+    end
+
+    def normalize_watcher_provided_opts(service_watcher_name, service_watcher_opts)
```

Comment:
we should either call this `opts` or `config`. seems like we settled on `config`